### ISSUE TITLE
Optimization - resolveExternal 

### ISSUE THREAD

**Issue by robfig** [bazelbuild/bazel-gazelle#867]
I noticed that about 40%-50% of Gazelle's time is spent blocked in `language/go.resolveExternal` (e.g. 7s-8s). It looks like that does single-threaded HTTP2 requests to get dependency module information. (We are using it in GOPATH mode, so it's calling repo.Root()) 

I was wondering if you had already thought about this and had any off-the-cuff ideas for how this might be sped up. If so, I'd be happy to take a look at implementing.

**Comment by jayconrod** [comment]
If it's possible, it's best to avoid requests in the first place. When Gazelle starts, it reads WORKSPACE and pre-populates `repo.RemoteCache` with the `go_repository` declarations it finds there. It will also read macros mentioned with `# gazelle:repository_macro` directives, and you can declare other repositories using `# gazelle:repository`. So if an import is already covered by a declared repository, Gazelle shouldn't have to go out to the network to resolve that.

Other than that, parts of fix-update.go can be parallelized, but it would involve refactoring major pieces like `repo.RemoteCache` and `resolve.RuleIndex` to be threadsafe. I'd rather not take PRs for that, since some refactoring work needs to be done first.

**Comment by robfig** [comment]
Ah, our `go_repository` declarations are not in WORKSPACE, so maybe that's why. They are actually generated by a Bazel rule - we're still using GLOCKFILE as the source of truth.

It sounds like a solution is to have a way to point Gazelle at our generated bzl file.. is that possible, or is that a flag you would support adding?

**Comment by robfig** [comment]
Er, it's a bit more complicated. We have a custom repository rule that takes the GLOCKFILE as an argument and returns the .bzl file content.

For example, 

```python
def _glock_repository_impl(ctx):
    res = ctx.execute([ctx.which("cat"), ctx.path(ctx.attr.glockfile)])
    if res.return_code != 0:
        fail(res.stderr)
        return
    ctx.file("BUILD")
    ctx.file("deps.bzl", content = glockfileBzl(res.stdout), executable = False)

glock_repository = repository_rule(
    implementation = _glock_repository_impl,
    local = True,
    attrs = {
        "glockfile": attr.label(
            mandatory = True,
            allow_single_file = True,
        ),
    },
)
```

and
```python
# glockfileBzl converts the contents of a glockfile to a Bazel deps.bzl file.
def glockfileBzl(glockfile):
    repo_decls = []
    for line in glockfile.splitlines():
        parts = line.split(' ')
        if len(parts) != 2:
            continue

        repo, hash = parts[0], parts[1]
        if repo == "cmd":
            continue

        repo_decls.append((bazelRepoName(repo), hash, repo))

    # NOTE: gax is v2 and uses a /v2 import path both internally and from a
    # couple dependencies. That import is resolved by module-aware Gazelle to
    # "@com_github_googleapis_gax_go_v2//", which does not work for us in GOPATH
    # mode. To address this, use the -known_import flag.
    #
    # TODO: Remove the build_extra_args when we've transitioned to modules.
    return """load("@bazel_gazelle//:deps.bzl", "go_repository")

def glock_dependencies():
""" + "\n".join(["""
    go_repository(
        name = "%s",
        commit = "%s",
        importpath = "%s",
        build_file_proto_mode = "disable",
        build_file_name = "BUILD.bazel",
        build_extra_args = ["-known_import=github.com/googleapis/gax-go","-known_import=github.com/cep21/circuit"],
    )""" % (repoName, hash, repo) for (repoName, hash, repo) in repo_decls])

```

I guess there's still be a generated file that could be passed as a flag to Gazelle.

**Comment by jayconrod** [comment]
In general, Gazelle won't invoke Bazel, so it can't read generated files or files in external workspaces.

However, it might already be possible to make this work.

There's a `-repo_config` flag that lets you point to a file instead of `WORKSPACE` where Gazelle should load information about known repositories. The `gazelle` rule could pass the generated file to Gazelle using that flag.

The format is the same as `WORKSPACE`, but the code isn't executed. The `# gazelle:repository` and `# gazelle:repository_macro` flags are recognized. Generating a bunch of `# gazelle:repository` directives would probably be the easiest thing.

```
# gazelle:repository go_repository name=org_golang_x_sync importpath=golang.org/x/sync
# gazelle:repository go_repository name=org_golang_x_sys importpath=golang.org/x/sys
# gazelle:repository go_repository name=org_golang_x_xerrors importpath=golang.org/x/xerrors
# gazelle:repository go_repository name=com_github_bazelbuild_buildtools importpath=github.com/bazelbuild/buildtools build_naming_convention=go_default_library
# gazelle:repository go_repository name=com_github_fsnotify_fsnotify importpath=github.com/fsnotify/fsnotify
# gazelle:repository go_repository name=com_github_pelletier_go_toml importpath=github.com/pelletier/go-toml
# gazelle:repository go_repository name=com_github_pmezard_go_difflib importpath=github.com/pmezard/go-difflib
# gazelle:repository go_repository name=com_github_bmatcuk_doublestar importpath=github.com/bmatcuk/doublestar
# gazelle:repository go_repository name=com_github_google_go_cmp importpath=github.com/google/go-cmp
```

**Comment by robfig** [comment]
I had a tough time passing the path to the generated file to Gazelle. 

To pass a generated file to `-repo_config`, I believe I need to change the `gazelle` bazel rule:

1. Update `//def.bzl` to add a `ctx.attr.label(name = "repo_config")` to the rule definition. 
2. In the implementation, if repo_config is provided, put that into runfiles
3. Update `//internal/gazelle.bash.in` to pass the value of `repo_config` through your `$find_runfile` function to turn it into an absolute path.

That allows me to do
```
gazelle(
  ...
  repo_config = "@glock//:deps.bzl",
)
```

I've uploaded a proof of concept that works. It might be possible to do this more cleanly; I was surprised how difficult it was to get the absoluate path to a generated file, and I saw no way to do it in our macro that wraps `gazelle()` or without editing the runner script.

What do you think?

https://github.com/bazelbuild/bazel-gazelle/pull/871
